src/diffusers/pipelines/unidiffuser/modeling_text_decoder.py

```diff
@@ -0,0 +1,220 @@
+        n_inner: Optional[int] = None,
+        activation_function: str = "gelu_new",
+        resid_pdrop: float = 0.1,
+        embd_pdrop: float = 0.1,
+        attn_pdrop: float = 0.1,
+        layer_norm_epsilon: float = 1e-5,
+        initializer_range: float = 0.02,
+        scale_attn_weights: bool = True,
+        use_cache: bool = True,
+        scale_attn_by_inverse_layer_idx: bool = False,
+        reorder_and_upcast_attn: bool = False,
+    ):
+        """
+        Text decoder model for a image-text [UniDiffuser](https://arxiv.org/pdf/2303.06555.pdf) model. This is used to
+        generate text from the UniDiffuser image-text embedding.
+
+        Parameters:
+            prefix_length (`int`):
+                Max number of prefix tokens that will be supplied to the model.
+            prefix_hidden_dim (`int`, *optional*):
+                Hidden dim of the MLP if we encode the prefix.
+            TODO: add GPT2 config args
+        """
+        super().__init__()
+
+        self.prefix_length = prefix_length
+        self.prefix_hidden_dim = prefix_hidden_dim
+        self.encode_prefix = (
+            nn.Linear(n_embd, self.prefix_hidden_dim) if self.prefix_hidden_dim is not None else nn.Identity()
+        )
+        self.decode_prefix = (
+            nn.Linear(self.prefix_hidden_dim, n_embd) if self.prefix_hidden_dim is not None else nn.Identity()
+        )
+
+        gpt_config = GPT2Config(
+            vocab_size=vocab_size,
+            n_positions=n_positions,
+            n_embd=n_embd,
+            n_layer=n_layer,
+            n_head=n_head,
+            n_inner=n_inner,
+            activation_function=activation_function,
+            resid_pdrop=resid_pdrop,
+            embd_pdrop=embd_pdrop,
+            attn_pdrop=attn_pdrop,
+            layer_norm_epsilon=layer_norm_epsilon,
+            initializer_range=initializer_range,
+            scale_attn_weights=scale_attn_weights,
+            use_cache=use_cache,
+            scale_attn_by_inverse_layer_idx=scale_attn_by_inverse_layer_idx,
+            reorder_and_upcast_attn=reorder_and_upcast_attn,
+        )
+        self.transformer = GPT2LMHeadModel(gpt_config)
+
+    def forward(
+        self,
+        tokens: torch.Tensor,
+        prefix: torch.Tensor,
+        mask: Optional[torch.Tensor] = None,
+        labels: Optional[torch.Tensor] = None,
+    ):
+        """
+        Args:
+            tokens (`torch.Tensor` of shape `(N, max_seq_len)`):
+                Text tokens to use for inference.
+            prefix (`torch.Tensor` of shape `(N, prefix_length, 768)`):
+                Prefix embedding to preprend to the embedded tokens.
+            mask (`torch.Tensor` of shape `(N, prefix_length + max_seq_len, 768)`, *optional*):
+                Attention mask for the prefix embedding.
+            labels (`torch.Tensor`, *optional*):
+                TODO
+        """
+        embedding_text = self.transformer.transformer.wte(tokens)
```

Comment:
Why `transformers.transformer`?

Reply:
I took the `forward(...)` method from the [original code](https://github.com/thu-ml/unidiffuser/blob/688f7b276396ac05931e8b014f2bc5e43239f3fd/libs/caption_decoder.py#L27-L37). Upon reviewing the code, I think the `forward` method was probably intended to do the following: the `tokens` argument is a sequence of input vocab token IDs for the `GPT2LMHeadModel`, while the `prefix` argument is the hidden state of another model (e.g. something like `transformers.modeling_outputs.BaseModelOutputWithPooling.last_hidden_state` of a `CLIPTextModel`). `prefix` then gets converted to an intermediate representation via `self.encode_prefix(...)` and then converted into the latent space of the GPT model via `self.decode_prefix(...)` (if they are being used). We then combine the embedding of `tokens` with the `prefix` embedding and then do a forward pass of the internal `GPT2LMHeadModel`.

I guess it's confusing currently because on lines 52-54 instead of using `n_embd` as the input dimension to `nn.Linear` we should instead have a new argument `prefix_inner_dim` and use that, e.g.

```python
self.encode_prefix = (
        nn.Linear(prefix_inner_dim, self.prefix_hidden_dim) if self.prefix_hidden_dim is not None else nn.Identity()
)
```

Reply:
Thanks a lot for explaining!

I guess we will know better once we start testing the code.